harness/tests/experiment/pytorch/test_deepspeed_trial.py

```diff
@@ -0,0 +1,651 @@
+                assert "loss" in metrics
+
+        controller = utils.make_trial_controller_from_trial_implementation(
+            trial_class=deepspeed_linear_model.LinearDeepSpeedTrial,
+            hparams=updated_hparams,
+            workloads=make_workloads(),
+            trial_seed=self.trial_seed,
+            expose_gpus=True,
+        )
+        controller.run()
+
+    def test_fail_manual_grad_acc_metrics(self) -> None:
+        updated_hparams = copy.deepcopy(self.hparams)
+        updated_hparams["test_fail_manual_grad_acc"] = True
+
+        def make_workloads() -> workload.Stream:
+            trainer = utils.TrainAndValidate()
+
+            yield from trainer.send(steps=10, validation_freq=10, train_batch_calls=1)
+            training_metrics, validation_metrics = trainer.result()
+
+            for metrics in validation_metrics:
+                assert "loss" in metrics
+
+        with pytest.raises(AssertionError):
+            controller = utils.make_trial_controller_from_trial_implementation(
+                trial_class=deepspeed_linear_model.LinearDeepSpeedTrial,
+                hparams=updated_hparams,
+                workloads=make_workloads(),
+                trial_seed=self.trial_seed,
+                expose_gpus=True,
+            )
+            controller.run()
+
+    def test_custom_dataloader(self) -> None:
+        updated_hparams = copy.deepcopy(self.hparams)
+        updated_hparams["test_manual_dataloader"] = True
+
+        def make_workloads() -> workload.Stream:
+            trainer = utils.TrainAndValidate()
+
+            yield from trainer.send(
+                steps=10,
+                validation_freq=10,
+                train_batch_calls=self.data_parallel_only_auto_train_batch_calls,
+            )
+            training_metrics, validation_metrics = trainer.result()
+
+            for metrics in validation_metrics:
+                assert "loss" in metrics
+
+        controller = utils.make_trial_controller_from_trial_implementation(
+            trial_class=deepspeed_linear_model.LinearDeepSpeedTrial,
+            hparams=updated_hparams,
+            workloads=make_workloads(),
+            trial_seed=self.trial_seed,
+            expose_gpus=True,
+        )
+        controller.run()
+
+    def test_fail_dataset_repro_check(self) -> None:
+        updated_hparams = copy.deepcopy(self.hparams)
+        updated_hparams["test_fail_dataset_repro_check"] = True
+
+        def make_workloads() -> workload.Stream:
+            trainer = utils.TrainAndValidate()
+
+            yield from trainer.send(
+                steps=10,
+                validation_freq=10,
+                train_batch_calls=self.data_parallel_only_auto_train_batch_calls,
+            )
+            training_metrics, validation_metrics = trainer.result()
+
+            for metrics in validation_metrics:
+                assert "loss" in metrics
+
+        with pytest.raises(RuntimeError):
```

Comment:
another `match=`

Reply:
I'm going to stop marking these.  There's a lot of them.

On the other hand, the reason there's so many is because you have great failure testing so nice work! :tada: